bittide-instances/src/Bittide/Instances/Hitl/HwCcTopologies.hs

```diff
@@ -488,54 +632,141 @@ hwCcTopologyTest refClkDiff sysClkDiff syncIn rxns rxps miso =
+    , disableReframing   = not cccEnableReframing
+    , rusty              = cccEnableRustySimulation
+    , waitTime           = fromEnum cccReframingWaitTime
+    , stopAfterStable    = Just
+                         $ natToNum @(PeriodToCycles Basic125 AllStablePeriod)
+    }
+
+  calibrateClockOffsets =
+    ( -- the name must be chosen such that this test is executed first
+      "0_calibrate_clock_offsets"
+    , ( toList $ imap (,) $ repeat @FpgaCount TestConfig
+          { fpgaEnabled       = True
+          , calibrate         = True
+          , stepSizeSelect    = commonStepSizeSelect
+          , initialClockShift = 0
+          , startupDelay      = 0
+          , mask              = maxBound
+          }
+      , defSimCfg
+          { mTopologyType  = Just $ Complete $ natToInteger @FpgaCount
+          , clockOffsets   = toList $ repeat @FpgaCount 0
+          , startupOffsets = toList $ repeat @FpgaCount 0
+          }
+      )
+    )
+
+  -- tests the given topology
+  tt ::
     forall n.
     (KnownNat n, n <= FpgaCount) =>
-    Topology n -> (TestName, (Probes TestConfig, SimConf))
-  testTopology t =
+    Vec n InitialClockShift ->
+    Vec n StartupDelay ->
+    Topology n ->
+    (TestName, (Probes TestConfig, SimConf))
+  tt clockShifts startupDelays t =
     ( fromString $ topologyName t
-    , ( toList (imap testData $ linkMasks @n t)
-         <> [ (fromInteger i, disabled)
-            | i <- [natToNum @n, natToNum @n + 1 .. natToNum @(FpgaCount - 1)]
-            ]
-      , def { mTopologyType = Just $ topologyType t
-            , simulationSamples = 1000
-            , simulationSteps = natToNum @(PeriodToCycles Basic125 (Seconds 60))
-            , stabilityMargin = snatToNum cccStabilityCheckerMargin
-            , stabilityFrameSize = snatToNum cccStabilityCheckerFramesize
-            , disableReframing = not $ cccEnableReframing
-            , rusty = cccEnableRustySimulation
-            , waitTime = fromEnum cccReframingWaitTime
-            , clockOffsets = toList $ repeat @n 0
-            , startupOffsets = toList $ repeat @n 0
-            , stopAfterStable =
-                Just $ natToNum @(PeriodToCycles Basic125 AllStablePeriod)
-            }
+    , ( toList ( zipWith4 testData indicesI clockShifts startupDelays
+               $ linkMasks @n t
+               )
+          <> [ (fromInteger i, disabled)
+             | let n = natToNum @n
+             , i <- [n, n + 1 .. natToNum @(FpgaCount - 1)]
+             ]
+      , let
+          clkPeriod :: Num a => a
+          clkPeriod = case clockControlConfig of
+            (_ :: ClockControlConfig dom a b c) ->
+              snatToNum (clockPeriod @dom)
+          stepSizeDiv = case commonStepSizeSelect of
+            PPB_1   -> 1_000_000
+            PPB_10  ->   100_000
+            PPB_100 ->    10_000
+            PPM_1   ->     1_000
+          conv s d x = ((/ d) . (s *)) . fromIntegral <$> toList x
```

Comment:
That name would be `turnEachElementOfTheVectorToNumThenMultiplyItByTheFirstArgAndThenDivideByTheSecond`, but that exceeds the 80 character line limit, so I had to be creative :upside_down_face:.

It's just a convenience function to avoid writing that "conversion" twice. We can also call it `f` instead, if you like?

Reply:
No it's Femtoseconds! All that conversion just happens because PPM is a stupid measure, which is also why NIST explicitly recommends **not** to use it (cf. https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.330-2019.pdf, p. 34). It's a relative measure, which is why it only makes sense in the context of something else (here the clock period). That is bad, because we are working with dynamic clocks, which makes the size of 1ppm change over time, if you would express it in Femtoseconds instead. This is also why we never can use PPM as a scale on the x-axis of of our plots, because it would not be independent of the y-axis (time).

To answer your question `conv` first converts the clock period (given in Picoseconds) to Femtoseconds and then applies the inverse of a 1000 times the PPM/PPB ratio or 1 (no ratio) given via the second argument (using an inverse scalar to avoid loss-of-precision).

To conclude, I recommend to completely elide PPM/PPB from our minds (and source code) and to use absolute measures instead, as it is also recommended by NIST.

Reply:
> That means it is relative to the input clock, right?

Yes, that is how I understood it as well.

The problem is that we measure the clock drift against the stable clock, which is why we also have to use something stable to label the y-axis. Using PPM instead, would be kind of confusing, as it is unclear which clock it relates to. The same applies to the integral over the number of applied FINC/FDECs, as the time portion, an FINC/FDEC changes the clock period, depends on the speed of a dynamic clock.

Reply:
> The problem is that we measure the clock drift against the stable clock, which is why we also have to use something stable to label the y-axis.

What do you mean with something "stable"? We fundamentally operate in PPM / relative land. If anything, current unit choice (fs) tricks the reader into believing there _is_ some absolute measurement we're taking.

Reply:
I don't think this is super relevant to the discussion, but the way I read the NIST report is that they're saying that million and billion can be confusing for non-English speakers (e.g., Dutch people say _biljoen_ for 10^12 instead of English's _billion_ which is 10^9). I'm not particularly convinced by that argument, we should speak English well enough. (And if this is about English long scale vs short scale I'm even less convinced.)

So if you mean that you want to ban PPM/PPB and replace it by some SI recommended, then: I don't think we should do that. PPM/PPB is prevalent in our industry. Both SkyWorks and Xilinx use it.

If you want to ban PPM/PPB because you want to switch to some absolute unit of time: as far as I can tell, that's just obfuscating the issue. From the POV from the reference clock (which I believe is now called "ideal frequency"?) you can only tell how much faster/slower the other clocks are _relative to_ it.

(I'm arguing from actual hardware BTW, in simulation we have access to absolute times of course.)

Reply:
> From the POV from the reference clock (which I believe is now called "ideal frequency"?) you can only tell how much faster/slower the other clocks are relative to it.

Right, but we consider the reference clock to be wall clock time, which is independent of the behavior of all the other clocks. Hence, this reference clock yields a time reference than can be considered absolute. In reality, this is not true for any clock source (due to Einstein's relativity) for sure, but it is a useful and standard abstraction to be utilized as soon as we are talking about wall clock time. 

Concluding to that: as long as we consider the reference clock to be stable and of known frequency, we always cancel it out from the equation and measure clock drift in absolute portions of time.

Reply:
I think we're talking past each other. Do you agree that for none of the clocks (controlled or non-controlled) we know how long a clock period exactly (`fs`) is?

Reply:
> but it is a useful and standard abstraction to be utilized as soon as we are talking about wall clock time.

I agree this is a good hack/workaround/assumption. But this seems to be a presentation thing, I don't see how it warrants completely getting rid of mentions of PPM?

Reply:
How about the mathematical precise version in the sense that `1ppm` is nothing else then $1^{-6}$ or `1e-6`? 

What I don't like about ppm is that most people use it like a unit and expect the reader to know what they mean. However, it's just a scale ratio. 

So I if tell you "it's 500000ppm!", then you basically always should ask "500000ppm of what?". Only if I tell you then that its "500000ppm of an apple" you know that I am actually talking about half an apple and not about clocks.

In that sense using $1^{-6}$ instead makes it immediately clear that it's just a scale ratio, just by looking at it. And and it also makes it clear that you still may have to add a unit or context in case you like to talk about something particular.

Reply:
Both should be accompanied by some context IMO. At least I hope we do for current mentions of PPM.

Reply:
> I don't think this is super relevant to the discussion, but the way I read the NIST report is that they're saying that million and billion can be confusing for non-English speakers

First of all, I think it'd be more relevant what the BIPM says rather than the NIST. Secondly, there are two arguments on that page of the NIST document. The body text notes that you shouldn't use ppb and ppt, but allows ppm. However, the footnote notes that ppm is not allowed in an SI-context.

I hope we generally fly by the SI, and will thus not _choose_ to use ppm ourselves. However, if the tooling and documentation around the clock boards all use ppm and ppb, it makes sense to use the same terminology in that part of the code that interfaces with it. It's still a voluntary choice, but it is very defensible.